internal/controller/management/serviceaccount/serviceaccounts.go

```diff
@@ -0,0 +1,229 @@
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+	"github.com/akuity/kargo/internal/logging"
+)
+
+const (
+	ComponentLabelKey = "app.kubernetes.io/component"
+	ControllerLabelValue = "controller"
+)
+
+type ReconcilerConfig struct {
+	KargoNamespace string `envconfig:"KARGO_NAMESPACE" required:"true"`
+}
+
+// ServiceAccountReconciler Reconciles for ServiceAccounts
+type ServiceAccountReconciler struct {
+	cfg ReconcilerConfig
+	client.Client
+}
+
+// SetupReconcilerWithManager initializes a reconciler
+// and registers that reconciler with the provided Manager.
+func SetupReconcilerWithManager(kargoMgr manager.Manager) error {
+	return ctrl.NewControllerManagedBy(kargoMgr).
+		For(&corev1.ServiceAccount{}).
+		WithEventFilter(predicate.Funcs{
+			CreateFunc: func(e event.CreateEvent) bool {
+				return hasControllerLabel(e.Object.(*corev1.ServiceAccount))
+			},
+			DeleteFunc: func(e event.DeleteEvent) bool {
+				return hasControllerLabel(e.Object.(*corev1.ServiceAccount))
+			},
+			UpdateFunc: func(e event.UpdateEvent) bool {
+				oldSA := e.ObjectOld.(*corev1.ServiceAccount)
+				newSA := e.ObjectNew.(*corev1.ServiceAccount)
+				return hasControllerLabel(oldSA) != hasControllerLabel(newSA)
+			},
+			GenericFunc: func(event.GenericEvent) bool {
+				return false
+			},
+		}).
+		Complete(&ServiceAccountReconciler{
+			Client:         kargoMgr.GetClient(),
+		})
+}
+
+// Reconcile handles the reconciliation logic for ServiceAccounts
+func (r *ServiceAccountReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+
+	logger := logging.LoggerFromContext(ctx).WithValues(
+		"serviceAccount", req.NamespacedName.Name,
+	)
+	ctx = logging.ContextWithLogger(ctx, logger)
+	logger.Debug("Reconciling ServiceAccounts")
+
+	// Fetch the ServiceAccount
+	sa := &corev1.ServiceAccount{}
+	if err := r.Get(ctx, req.NamespacedName, sa); err != nil {
+		if kubeerr.IsNotFound(err) {
+			logger.Debug("ServiceAccount not found, deleting role bindings for all projects")
+			// If not found it means the ServiceAccount with the 
+			// controller label was deleted, hence remove Controller
+			// Permissions for the ServiceAccount.
+			return ctrl.Result{}, r.deleteRoleBindingsForAllProjects(ctx, sa)
+		}
+		logger.Error(err, "Failed to get ServiceAccount")
+		return ctrl.Result{}, err
+	}
+
+	if controllerutil.AddFinalizer(sa, kargoapi.FinalizerName) {
+		if err := r.Update(ctx, sa); err != nil {
+			return ctrl.Result{}, err
+		}
+		logger.Debug("Added finalizer to ServiceAccount", "serviceaccount", sa.Name)
+		return ctrl.Result{}, nil
+	}
```

Comment:
We shouldn't be doing this unconditionally. As it stand now, you have the possibility of adding the finalizer here only to remove it a few lines down from here.